subschema composability rules

yaacovCR opened the pull request:
Comments welcome!

This is just an initial version, and has much room for improvement.

yaacovCR committed:
initial Subschema Composability riffs

WIP

See: https://github.com/graphql/composite-schemas-wg/discussions/35

yaacovCR committed:
language edits

yaacovCR committed:
use merging for types more consistently rather than composing

yaacovCR committed:
remove references to partial composability

yaacovCR committed:
Add case where enum types contain equivalent value sets

yaacovCR committed:
add some initial thoughts about input objects and fields

JohnStarich commented on rfcs/SubschemaComposability.md:
```diff
@@ -0,0 +1,74 @@
+
+1. Overlapping input object fields can be merged as long as the fields define the same type, allowing for variations in nullability.
+2. If any of the field types contain non-nullable portions, the field types must be merged as non-nullable for that portion.
+3. TODO: discuss default values.
```
`3. ... default values` is interesting. Is a default value normally filled in by an executor? I suppose this could vary across implementations.

yaacovCR replied:
Default values are encoded by subschemas and substituted by the executor if no value is sent. If there is a discrepancy between subschemas with regards to the default, ie whether one is defined or what the value is, I think a merge should still succeed, and I basically agree that we don't really need to specify whether/how the composite schema should define a default. :)

yaacovCR replied:
I do not have direct experience with Nautilus but from the docs it looks like types are composed if they have the same name, and that cross service runtime resolution is by using the Node interface and node root query field.

It does not say what happens if object types of the same name do not implement Node or if the node root query field is absent. Presumably, composition will fail or the types must be namespaced.

Assuming the latter, the use of the Node interface does not seem to be a separate way to maintain composability per se; rather, the composition proceeds by name spacing by default unless type composability is possible. This is certainly a sane default rather than simply failing.

I am making a lot of assumptions from the docs I could find, please let me know how/if I have misunderstood how Nautilus works. :)

JohnStarich reviewed (commented):
@yaacovCR This is great. 👏 I've also been thinking about mergability and a common merging spec as a primary concern.

I've added some of my riffs as suggestions to your riffs 🙂

yaacovCR commented on rfcs/SubschemaComposability.md:
```diff
@@ -0,0 +1,74 @@
+   - When using `Single-service composite schema execution`, the enum types can be merged as a union of values from all subschemas.
+   - When using `Subschemas as remote GraphQL services`
+     - If the enum types are used only in output types, the enum types can be merged as a union of values from all subschemas.
+     - If the enum types are used in input types, the enum types can be merged if and only if there is a way to translate, for any given subschema, every enum value that is defined only in the composite schema to an enum value defined for that subschema.
```
I briefly posted a version that allowed intersections of enums, but I think that unfortunately we have to fail the composition altogether.

For example, consider two scenarios:

Perhaps a field within a subschema defines an argument called order of enum type OrderEnum with values ASCENDING and DESCENDING. If DESCENDING is filtered out from the composed schema, the composed schema has fewer capabilities than at least one of the subschemas, but all data represented by the subschema is accessible.

Consider another scenario in which a field has an argument called filter of enum type FilterEnum with values ItemTypeOne, ItemTypeTwo etc that are meant to be an exhaustive list of the possible item types returned by this field. If values for this enum are filtered from the enum in the composite schema, then the functionality of the filter argument no longer makes sense, and the whole argument should probably not be included in the composite schema.

Seeing a subset would be surprising in both cases, but I think it would be preferable in the latter case to fail the merge. Unfortunately, I don't see a straightforward way to delineate between the two cases without human knowledge of the behavior of the argument.

JohnStarich replied:
I didn't quite follow the first example, but your second example makes perfect sense. 👍 
We simply can't have an output enum value which isn't already described by the full schema. Bummer!

yaacovCR replied:
The problem of remote subschema breaking changes bubbling up to the composed schema is pretty serious — but is it more serious than any remote service breaking change would be for a non-composed GraphQL service?

In any event, I am not clear how subschema metadata would help us navigate the challenge. If a subschema removes a field, for example, what kind of metadata could help?

If a subschema adds a field that clashes with another subschema, presumably the composed schema has no knowledge of the change, so the clash will not actually cause a runtime error.

Standardizing subschema metadata seems useful to me, although I was mostly thinking about in terms of nomenclature for how to perform cross-service resolution at runtime, I.e. key directives…

yaacovCR replied:
Quick sanity check: did you mean input enum value?

yaacovCR commented on rfcs/SubschemaComposability.md:
```diff
@@ -0,0 +1,74 @@
+# Input Object Types
+
+1. Overlapping input object types can be merged, as long as the overlapping input fields can be merged.
+2. If some of the types define fields not defined by the other types, the fields of these input objects may no longer affect the execution of portions of the subschema, in ways that may be unpredictable for users of the composite schema. Although clear documentation may ameliorate this issue, overall, the user experience may suffer. 
```
I think the merge should probably fail in this case as well.

yaacovCR committed:
remove use of term "type merging"

to drop the number of used terms/simply, we can just use type composition

yaacovCR committed:
significant rewrite

1. separate composability rules by execution strategy.

I am not sure if these should be separate in the final document, but separating helps me reason about them better within this draft stage.

2. remove discussion of runtime translation from composability rules

thats basically mixing the idea of "transforms" with composability.

For example, if there is a way of translating a missing enum value for a given subschema into an enum value that is present, we are basically transforming that subschema into a subschema that has the missing enum value by using a transformRequest method

Instead, let's make the composability rules strict and discuss separately the idea of transforms

yaacovCR committed:
standardize

yaacovCR committed:
Add objects, interfaces, unions

yaacovCR committed:
switch order vis a vis execution strategies

yaacovCR committed:
prettier

yaacovCR committed:
unify

yaacovCR commented:
I have marked this PR as no longer DRAFT, as I have added section regarding all types. But it would certainly benefit from additional input and further revision. Comments are welcome.

JohnStarich commented on rfcs/SubschemaComposability.md:
```diff
@@ -0,0 +1,239 @@
+The below RFC examines composability separately for the above two forms of
+composite schema execution.
+
+# Techniques for Maintaining Composability
```
We may want to also consider developing some new techniques for maintaining composability, too. I recall we discussed the issue where schemas could be composed but the schema owners have no idea it is being composed, which creates problems when it becomes incompatible with other subschemas in the future.

One strategy might be to explicitly "opt-in" to composition. In other words, the schema contains metadata which says "This schema is allowed to be a subschema."

Going a step further, the opt-in metadata (directive?) could indicate a specific composition strategy. For example, say I wanted to provide my schema as-is. However, if someone wanted to use my schema as a subschema, then I could include a directive that says "if you compose this, then use this namespace/transform."

Another option might be simply: The composer (e.g. registry) knows which subschemas should be composed and all are guaranteed to be opt-in. This isn't always true but it may be a decent, practical solution.

This might be better suited in the discussions tab, however I am curious what our initial thoughts are. Maybe it'd be worth discussing this week in the WG, who knows 🙂

yaacovCR replied:
The problem of remote subschema breaking changes bubbling up to the composed schema is pretty serious — but is it more serious than any remote service breaking change would be for a non-composed GraphQL service?

In any event, I am not clear how subschema metadata would help us navigate the challenge. If a subschema removes a field, for example, what kind of metadata could help?

If a subschema adds a field that clashes with another subschema, presumably the composed schema has no knowledge of the change, so the clash will not actually cause a runtime error.

Standardizing subschema metadata seems useful to me, although I was mostly thinking about in terms of nomenclature for how to perform cross-service resolution at runtime, I.e. key directives…

yaacovCR replied:
I should also note that robustly maintaining composability across schema changes in my mind is the job of the composition service and may in large part consist of managing a schema registry for the subschemas.

Metadata that might be useful could be a schema version, or a hash of the SDL.

JohnStarich replied:
All good points. After thinking it through a bit more, the metadata may not help as much. The hope there was to standardize how a subschema wanted to be composed, but that may only be useful when trying to isolate subschemas from one another.

The backward compatibility issue is especially pertinent. Maybe we can kick the can for now, and in the future some new guidelines could recommend ways to handle those issues.

>I should also note that robustly maintaining composability across schema changes in my mind is the job of the composition service and may in large part consist of managing a schema registry for the subschemas.

I agree, the composition service should be responsible (at least in part) for safe and back-compatible merges. However, it seems how this should be done is still up in the air.

JohnStarich commented on rfcs/SubschemaComposability.md:
```diff
@@ -0,0 +1,239 @@
+subschemas as remote GraphQL services, the subschema transformation may also
+require runtime transformation of the requests to the subservices and/or the
+subservice results.
+
```
One composition technique I found useful for my team is employed by [Nautilus](https://github.com/nautilus/gateway). It composes subschema object types if and only if they opt-in. They opt-in by implementing a shared interface with only an `id: ID!` field. It appears to be unique in that it isn't specifying directives or special syntax.

The spec for "global object identification" lends itself well to schema merging for my team's needs. This solved many of our schema composition problems in a very simple way, i.e. subschemas did not need to go out of their way to opt-in to composition for objects.

This may even be in a separate category from Namespaces or Transforms. Curious what you think there.

obi1kenobi replied:
Personally, I think it's a great technique when it's available, but it's not always an option.

For example, in my team's usecase [at Kensho](https://blog.kensho.com/database-agnostic-querying-is-unavoidable-at-scale-18895f6df2f0), we could not have used this approach because we didn't "own" the underlying data sources, and many of them didn't even have a single (or any) primary key that we could reference. There was data with composite keys, and data that was generated on demand and had no meaningful natural key of any kind whatsoever. Additionally, for the most part the underlying services weren't aware they were being queried through GraphQL and a metadata-driven federated execution system, and could not meaningfully have opted into federation.

I completely agree that opt-in is great and should be used whenever possible. However, mandating opt-in behavior in the standard would disallow use cases like the one I described and linked to, which would be a shame and would likely lead to more non-standard extensions being developed and used.

JohnStarich replied:
@obi1kenobi You make some good points about the tradeoffs of using `Node` as an opt-in. This was mainly a suggestion for an additional "in the wild" strategy for merging, below namespacing and transforms.

JohnStarich reviewed (commented):
I had some more thoughts on "what does it mean to be composable?". This feels like it could be a whole separate topic, so I'm happy to move this to a discussion too.

gkesler commented on rfcs/SubschemaComposability.md:
```diff
@@ -0,0 +1,239 @@
+   composite schema will yield request(s) against spec-compliant (or even
+   non-spec compliant) GraphQL subservices that are then merged into a single
+   request.
+2. `Single-service composite schema execution`: Often, especially when utilizing
```
I think this needs a bit of clarification. For instance if one of the services has a graphql endpoint, then all requests would be formatted as graphql requests. In such case this case wouldn't be different from the first case.
Maybe this case could be better understood if it has "non-GrapgQL services" somewhere.

obi1kenobi commented:
This is a great doc! I don't find the GitHub PR interface great for reviewing Markdown, so I wrote up a few thoughts in the discussion thread:
https://github.com/graphql/composite-schemas-wg/discussions/35#discussioncomment-3400447

benjie commented:
Minor markdown styling comment: please use

```
`backticksForCode`
```

and

```
*italics for references*
```

i.e. change
```
`Subschemas as remote GraphQL services`
```

to

```
*Subschemas as remote GraphQL services*
```

This follows the pattern used by spec-md.

michaelstaib commented on rfcs/SubschemaComposability.md:
```diff
@@ -0,0 +1,239 @@
+
+For a given set of subschemas, it may be possible to generate a valid
+`composite schema` that contains all GraphQL elements of all of the individual
+subschemas. This set of subchemas is then considered to be `composable`.
```
```suggestion
subschemas. This set of subschemas is then considered to be `composable`.
```

michaelstaib commented on rfcs/SubschemaComposability.md:
```diff
@@ -0,0 +1,239 @@
+   build process prior to system start-up.
+2. "Run-time" composition in which the composite schema is created dynamically
+   at system start-up.
+3. "Dynamic" composition , in that subschemas are polled by a managing service
```
```suggestion
3. "Dynamic" composition, in that subschemas are polled by a managing service
```

yaacovCR committed:
Feedback

Thanks @michaelstaib

Co-authored-by: Michael Staib <michael@chillicream.com>

yaacovCR committed:
Feedback

Thanks @michaelstaib

Co-authored-by: Michael Staib <michael@chillicream.com>

yaacovCR commented:
> This follows the pattern used by spec-md.

@benjie any styling for definitions like for “composability” ?

benjie commented:
@yaacovCR Indeed; if you're _actually_ using spec-md then the paragraph that defines a term should start with `:: ` and the term should be the first italicised term. Then references to it should use italics. Here's an example of `request` being defined in the spec:

https://github.com/graphql/graphql-spec/pull/949/files

benjie commented on rfcs/SubschemaComposability.md:
```diff
@@ -0,0 +1,239 @@
+
+1. Overlapping enum types where the types define identical sets of values can
+   always be composed, as they are identical in all subschemas.
+2. Subschemas with overlapping enum types where the disparate types define
+   different value sets are only sometimes composable.
```
```suggestion
2. Subschemas with overlapping enum types where the disparate types define
   different value sets are sometimes composable.
```

benjie commented on rfcs/SubschemaComposability.md:
```diff
@@ -0,0 +1,239 @@
+
+## Interface Types
+
+1. Overlapping interface types cannot be composed if any overlapping fields
+   cannot be composed, see below.
```
May need a comment on the interfaces that the interface implements.

benjie commented on rfcs/SubschemaComposability.md:
```diff
@@ -0,0 +1,239 @@
+
+## Object Types
+
+1. Overlapping object types cannot be composed if any overlapping fields cannot
+   be composed, see below.
```
Needs a comment on the interfaces the object type implements too.

benjie reviewed (changes requested):
Should we go ahead and merge this since it's an RFC doc?

JohnStarich commented:
During our call today we outlined where we're going, and sounds like composability rules are on the horizon again.

@michaelstaib @martijnwalraven We could use Yaacov's PR here as a discussion starter.

trkohler commented:
in the rfc there were several times mentioned ways to influence schema objects composability 

>  a method of composition may be devised in which elements in
different subschemas with different names are mapped to the same element in the
composite schema via some set of guiding metadata.

As far as I understood, it must be some directives, which allow merging or do merging implicitly? May small paragraph on directives which supply merging will be useful there?

(please just ignore my input, if it doesn't help!)

smyrick commented:
@michaelstaib @PascalSenn @martijnwalraven We can probably close this and https://github.com/graphql/composite-schemas-wg/pull/34/files since we have moved the spec docs to https://github.com/graphql/composite-schemas-spec
